### ISSUE TITLE
Sdkman cannot be used with JVM in codespaces

### ISSUE THREAD

**Issue by dsyer** [devcontainers/images#187]
Sdkman is installed by default in the base images, which is potentially quite useful, but it turns out that the base container/VM is doing something special with `${HOME}/.java` which conflicts with it.

In codespaces:

```
$ echo $PATH | tr : '\n' | grep java
/home/codespace/.java/current/bin
/usr/local/sdkman/candidates/java/current/bin
```

In a devcontainer created locally (just with default settings for Java):

```
$ echo $PATH | tr : '\n' | grep java
/usr/local/sdkman/candidates/java/current/bin
/usr/local/sdkman/candidates/java/current/bin
```

The net result is that if you `sdk install java xxx` and `sdk use java xxx` it isn't actually the default java because of the way the `PATH` is set up in the codespace.

**Comment by samruddhikhandale** [comment]
Hi 👋 

I tried to reproduce in a codespace built with `mcr.microsoft.com/devcontainers/universal` image. I was able to successfully set a new default java version.

Attempt 1  - `Do you want java 19.0.1-oracle to be set as default? (Y/n): y`
![image](https://user-images.githubusercontent.com/24955913/199573537-e3a63d0b-7dff-4257-bd27-b4b69030658b.png)


Attempt 2 
![image](https://user-images.githubusercontent.com/24955913/199574484-e712dd9a-5c07-4949-83ad-919138843747.png)


@dsyer could you use the `sdk default java xxx` command instead? It worked for me.

**Comment by samruddhikhandale** [comment]
ℹ️ We are creating a symlink `/home/codespace/.java/current -> /usr/local/sdkman/candidates/java/current`. Hence, whichever java version resides in `/usr/local/sdkman/candidates/java/current` should get reflected with PATH `/home/codespace/.java/current`

**Comment by dsyer** [comment]
You must be using some different configuration than the default (for me):

```
$ ls -l ~/.java/current
lrwxrwxrwx 1 codespace codespace 17 Oct 20 13:07 /home/codespace/.java/current -> /opt/java/11.0.14
$ sdk use java 17.0.5-tem 

Using java version 17.0.5-tem in this shell.
$ ls -l ~/.java/current
lrwxrwxrwx 1 codespace codespace 17 Oct 20 13:07 /home/codespace/.java/current -> /opt/java/11.0.14
$ java -version
openjdk version "11.0.14.1" 2022-02-08 LTS
OpenJDK Runtime Environment Microsoft-31205 (build 11.0.14.1+1-LTS)
OpenJDK 64-Bit Server VM Microsoft-31205 (build 11.0.14.1+1-LTS, mixed mode)
```

```
$ sdk default java 17.0.5-tem 

Default java version set to 17.0.5-tem
$ ls -l ~/.java/current
lrwxrwxrwx 1 codespace codespace 17 Oct 20 13:07 /home/codespace/.java/current -> /opt/java/11.0.14
```

**Comment by samruddhikhandale** [comment]
In my codespace (with default `universal` image), I have the following configuration

```
ls -l ~/.java/current
lrwxrwxrwx 1 codespace codespace 41 Nov  2 18:45 /home/codespace/.java/current -> /usr/local/sdkman/candidates/java/current
```

This is how the symlink is setup in the image [here](https://github.com/devcontainers/images/blob/main/src/universal/.devcontainer/local-features/setup-user/install.sh#L48) which guarantees the `default` set by `sdk` should reflect with the one in a codespace. 😄 

@dsyer Could you provide me your repro steps? Also, if you could run ` devcontainer-info` command in the codespace, it will provide image details for your codespace.

**Comment by dsyer** [comment]
Repro steps for me really simple: open a Java (Maven) project in codespaces and run the commands above.

```
$ devcontainer-info

Development container image information

- Image version: 2.0.20
- Definition ID: universal
- Source code repository: https://github.com/devcontainers/images
- Source code release/branch: v0.2.18
- Timestamp: Tue, 18 Oct 2022 22:47:47 GMT

More info: https://github.com/devcontainers/images/tree/main/src/universal/history/2.0.20.md
```

**Comment by schrum2** [comment]
Sorry to butt in, but I just want to confirm that I'm experiencing the same issue I believe. I was encouraged to follow this thread after submitting this issue, which is sort of partially fixed: https://github.com/orgs/community/discussions/37051

Basically, I make a brand new codespace, execute `gradle build` (instead of Maven), and then try running the jar file with `java` but get this error because the default java does not match the one that gradle is using:
```
java.lang.UnsupportedClassVersionError: bio/Test has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0
```
My work around is to run `$JAVA_HOME/bin/java` instead, but I shouldn't need to do this.

I can provide access to a minimal repo where this issue can be reproduced if that would help. In any case, running the `devcontainer-info` on my codespace gives similar output, though the image version is slightly different:
```
$ devcontainer-info

Development container image information

- Image version: 2.0.22
- Definition ID: universal
- Source code repository: https://github.com/devcontainers/images
- Source code release/branch: v0.2.21
- Timestamp: Thu, 27 Oct 2022 17:04:27 GMT

More info: https://github.com/devcontainers/images/tree/main/src/universal/history/2.0.22.md
```

**Comment by dsyer** [comment]
I find that `rm ~/.java/current` fixes things for me (sdkman takes care of the JAVA_HOME much better). Kind of a pain to have to remember to do that every time, so it really does seem like a bug to me.

**Comment by dsyer** [comment]
Related question: how do you get a base container to bump its revision (@schrum2 has a newer version than me)?

**Comment by samruddhikhandale** [comment]
> Related question: how do you get a base container to bump its revision 

If you create a new codespace, you should be able to get the latest version deployed by `codespaces` 😄 
If you rebuild (full) your codespace, then you should get the latest image deployed by this repo.

**Comment by samruddhikhandale** [comment]
When `oryx` detects a `Java` Project, it tries to install the following tools -
- Java
    - It installs it's own version of Java in two cases - 1) if java is not found 2) If the detected java version is not supported by oryx ([list](https://github.com/microsoft/Oryx/blob/main/doc/supportedPlatformVersions.md#java)) 
    - It installs java from source instead of using sdkman
- Maven

The default image in Codespaces (ie universal) installs two versions of `Java` along with `Maven` & `Gradle`. I think the auto-installation of tools by oryx is not helpful for `Java` projects.

@jkeech @Chuxel Any converns with disabling auto-detection/installation by oryx for `Java` projects? This would help fix the user issue mentioned in OP.

ℹ️  We can simply disable by setting `DISABLE_JAVA_BUILD` variable ([ref](https://github.com/microsoft/Oryx/blob/main/src/BuildScriptGeneratorCli/SettingsKeys.cs#L28)) .

Another alternative to fix this issue, is to modify the oryx tool & allow it to install java using `sdkman`. However, it could take a while to modify the upstream logic & we'd need to contact the oryx team to ask for permission.

**Comment by Chuxel** [comment]
If you use SDKMAN as the JDK path, you can add the Oryx install location as the default by adding to SDKMAN as something like "system" or "oryx" and set it to be the default.

That said, SDKMAN should be getting sourced though in bashrc or zshrc which should overwrite whatever is already in the PATH.  Is that not happening?

Something must be incorrectly placing the $HOME/.java path ahead of SDKMAN which hasn't happened historically.

**Comment by samruddhikhandale** [comment]
`Universal` image uses sdkman to install java. It then adds a sym link as follows 👇  -

![image](https://user-images.githubusercontent.com/24955913/202556227-a87c77ab-44f5-4184-81db-35b0f6e17245.png)

Even if the user installs another version of java with SDKMAN and sets it as default, then it is correctly reflected. (as `$HOME/.java` is always placed infront of SDKMAN in PATH & it works perfectly)

However, for a java project - oryx installs another version of java to `/opt/java/*` (from source) and links that to `$HOME/.java`. As `$HOME/.java` is ahead of SDKMAN in PATH, the detected java version is always the one in `$HOME/.java` which is sym linked to `/opt/java/*.

Hence, if the user tries to install & set another java version with sdkman, then it is reflected in `/usr/local/sdkman/candidates/java/current` but $HOME/.java is in the front, so it's not reflected when user runs `java --version`.

Hence, I am proposing to disable auto-installation of java by oryx (explanation [here](https://github.com/devcontainers/images/issues/187#issuecomment-1319126832)) or we could simply add `SDKMAN` in front of the PATH (however, the creation log would be contradictory to what is currently present).

**Comment by Chuxel** [comment]
Yeah the path is backwards from what it has been historically, and Oryx was what added the versions in the image. The only thing that was done with SDKMAN was adding the Oryx location to the list of available SDKMAN candidates to allow you to switch back if you want to. 

FWIW - Node, Ruby, etc would have this same problem since what is in the version manager (nvm, rvm) should always win.

**Comment by samruddhikhandale** [comment]
PRs to fix this bug are now in review.

- https://github.com/devcontainers/images/pull/219
- https://github.com/microsoft/Oryx/pull/1692

**Comment by samruddhikhandale** [comment]
Released a fix with [universal:2.1.1](https://github.com/devcontainers/images/blob/main/src/universal/history/2.1.1.md)

![image](https://user-images.githubusercontent.com/24955913/204855425-ff97c813-f7f9-44e8-8272-0af8f2acafec.png)

Now, waiting on a Codespaces deployment.

**Comment by samruddhikhandale** [comment]
Closed as fixed this bug in a Codespace. Feel free to re-open if the issue resurfaces.
